lib/reqTemplate.js

```diff
@@ -75,16 +65,53 @@ function splitAndPrepareTAsseblyTemplate(templateSpec) {
 }
 
 /**
- * Compiles a template into a function
- *
- * @param templateSpec template string (e.g. '{$.request.body.a}')
- * @param reqPart part of the request to lookup local properties in (e.g. headers, body etc.)
- * @param setValue callback for a result of template evaluation
- * @returns {Function} a template resolver for the provided template
+ * Creates a template resolver functuons for URI part of the spec
+ * @param {object} spec a root request spec object
+ * @returns {Function} a template resolver which should be applied to resolve URI
  */
-function compileTemplate(templateSpec, setValue, reqPart) {
+function createURIResolver(spec) {
+    if (/^\{[^\{}]+}$/.test(spec.uri) || /\{\$\$?\..+}/.test(spec.uri)) {
+        var tassemblyTemplate = splitAndPrepareTAsseblyTemplate(spec.uri);
+        var resolver = compileTAssembly(tassemblyTemplate, 'params');
+        return function(context) {
+            var value = resolver(context);
+            if (value.constructor !== URI) {
+                value = new URI(value, {}, false);
+            }
+            return value;
+        };
+    } else if (/^(?:https?:\/\/)?\{[^\/]+}\//.test(spec.uri)) {
+        // The host is templated - replace it with TAssembly and use URI.expand for path templates
+        var hostTemplate = /^((?:https?:\/\/)?\{[^\/]+}\/)/.exec(spec.uri)[1];
+        var hostTassembly = splitAndPrepareTAsseblyTemplate(hostTemplate);
+        var hostResolver = compileTAssembly(hostTassembly, 'params');
+        var path = spec.uri.substr(hostTemplate.length);
+        var pathTemplate = new URI('/' + path, {}, true);
+        return function(context) {
+            var newHost = hostResolver(context);
+            // FIXME: Support references to other parts of the request.
+            // params['$'] = context.rm;
+            var newUri = pathTemplate.expand(context.rm.request.params);
+            newUri.urlObj = url.parse(newHost + path);
+            return newUri;
+        };
+    } else {
+        return (function(uri) {
+            var uriTemplate = new URI(uri, {}, true);
+            return function(context) {
+                return uriTemplate.expand(context.rm.request.params);
+            };
+        })(spec.uri);
```

Comment:
This self-invoking closure was added to avoid seeing original spec's elements rewrites, so as now it's cloned, I'm quite sure it's not needed, but not 100% sure.